File: src/Dez4/P2.java
package Dez4;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class P2 {

    static ArrayList<char[]> list = new ArrayList<>();

    static int[] dirX = {-1, -1, 1, 1};
    static int[] dirY = {-1, 1, 1, -1};

    public static void main(String[] args) throws IOException {
        readFile();
        System.out.println(countXMas());
    }

    static void readFile() throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader("/Users/gd860/IdeaProjects/AoC24/src/Dez4/P1.txt"))) {
            String line;
            while ((line = reader.readLine()) != null) {
                list.add(line.toCharArray());
            }
        }
    }

    static int countXMas() {
        int result = 0;
        for (int x = 0; x < list.size(); x++) {
            for (int y = 0; y < list.get(x).length; y++) {
                if (checkDir(x, y)) {
                    result++;
                }
            }
        }
        return result;
    }


    static boolean checkDir(int x, int y) {
        boolean isValid = false;
        if (list.get(x)[y] == 'A') {
            if (!isValidPos(x + 1, y + 1)) {
                return false;
            }
            if (!isValidPos(x - 1, y + 1)) {
                return false;
            }
            if (!isValidPos(x + 1, y - 1)) {
                return false;
            }
            if (!isValidPos(x - 1, y - 1)) {
                return false;
            }

            char lastChar = 'Q';
            int countM = 0;
            for (int i = 0; i < 4; i++) {
                char current = list.get(x + dirX[i])[y + dirY[i]];
                if (current == 'M' || current == 'S') {
                    if (lastChar == current) {
                        isValid = true;
                    }
                    lastChar = current;
                    if (current == 'M') {
                        countM++;
                    }
                } else {
                    return false;
                }
            }
            return isValid && countM == 2;
        }
        return false;
    }

    static boolean isValidPos(int x, int y) {
        return x >= 0 && x < list.size() && y >= 0 && y < list.get(0).length;
    }
}

File: src/Dez6/P2.kt
package Dez6

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException

private fun main() {
    readFile()
    findGuard()
}

private var list = ArrayList<CharArray>()

private var dir = 0;

@Throws(IOException::class)
private fun readFile() {
    val file = File("/Users/gd860/IdeaProjects/AoC24/src/Dez6/P1.txt")
    BufferedReader(FileReader(file)).use { reader ->
        var line: String?
        while (reader.readLine().also { line = it } != null) {
            line?.let { list.add(it.toCharArray()) }
        }
    }
}

private fun findGuard() {
    var posX = 0;
    var posY = 0;
    for (x in list.indices) {
        for (y in list.indices) {
            if (list[x][y] == '^') {

                posX = x
                posY = y
            }
        }
    }
    calcPath(posX, posY)
}

private fun calcPath(x: Int, y: Int) {
    var posX = x;
    var posY = y;

    var checkPosX = x;
    var checkPosY = y;

    var loops = 0;

    checkPosX--

    while (isValidPos(checkPosX, checkPosY)) {
        if (list[checkPosX][checkPosY] == '.') {
            posX = checkPosX
            posY = checkPosY
            list[posX][posY] = getSymbolForDir()
        } else if (list[checkPosX][checkPosY] == '#') {
            extendPathBack(posX, posY)
            if (dir != 3) {
                dir++;
            } else {
                dir = 0
            }
            checkPosX = posX
            checkPosY = posY
        }

        when (dir) {
            0 -> {
                checkPosX--
            }

            1 -> {
                checkPosY++
            }

            2 -> {
                checkPosX++
            }

            3 -> {
                checkPosY--
            }
        }
    }
    for (row in list) {
        for (col in row) {
            print(col)
        }
        println()
    }

    println()
    println()
    println()

    checkPosX = x;
    checkPosY = y;

    dir = 0

    while (isValidPos(checkPosX, checkPosY)) {
        if (list[checkPosX][checkPosY] == '.' || list[checkPosX][checkPosY] == getSymbolForDir()) {
            posX = checkPosX
            posY = checkPosY
            list[posX][posY] = getSymbolForDir()
        } else if (list[checkPosX][checkPosY] == getNextSymbolForDir()) {
            posX = checkPosX
            posY = checkPosY
            list[checkPosX][checkPosY] = 'O'
            loops++
        } else if (list[checkPosX][checkPosY] == '#') {
            extendPathBack(posX, posY)
            if (dir != 3) {
                dir++;
            } else {
                dir = 0
            }
            checkPosX = posX
            checkPosY = posY
        }

        when (dir) {
            0 -> {
                checkPosX--
            }

            1 -> {
                checkPosY++
            }

            2 -> {
                checkPosX++
            }

            3 -> {
                checkPosY--
            }
        }
    }
    for (row in list) {
        for (col in row) {
            print(col)
        }
        println()
    }

    print(loops)
}

private fun isValidPos(x: Int, y: Int): Boolean {
    return x >= 0 && x < list.size && y >= 0 && y < list[0].size
}

private fun getSymbolForDir(): Char {
    when (dir) {
        0 -> {
            return '^'
        }

        1 -> {
            return '>'
        }

        2 -> {
            return 'v'
        }

        3 -> {
            return '<'
        }
    }
    print("error")
    return 'X'
}

private fun getNextSymbolForDir(): Char {
    when (dir) {
        0 -> {
            return '>'
        }

        1 -> {
            return 'v'
        }

        2 -> {
            return '<'
        }

        3 -> {
            return '^'
        }
    }
    print("error")
    return 'X'
}

private fun extendPathBack(x: Int, y: Int) {
    var posX = x;
    var posY = y;

    while (true) {
        when (dir) {
            0 -> {
                posX++
            }

            1 -> {
                posY--
            }

            2 -> {
                posX--
            }

            3 -> {
                posY++
            }
        }
        if (!isValidPos(posX, posY)) {
            break
        } else if (list[posX][posY] == '#') {
            break
        } else {
            list[posX][posY] = getSymbolForDir()

        }
    }

}